components/script_plugins/unrooted_must_root.rs

```diff
@@ -139,97 +114,148 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnrootedPass {
+impl<'a, 'b, 'tcx> UnrootedCx<'a, 'b, 'tcx> {
+    fn is_unrooted_ty(&self, ty: &ty::TyS, in_new_function: bool) -> bool {
+        let mut ret = false;
+        let cx = self.late_cx;
+        ty.maybe_walk(|t| {
+            match t.sty {
+                ty::TyAdt(did, _) => {
+                    if cx.tcx.has_attr(did.did, "must_root") {
+                        ret = true;
+                        false
+                    } else if cx.tcx.has_attr(did.did, "allow_unrooted_interior") {
+                        false
+                    } else if match_def_path(cx, did.did, &["core", "cell", "Ref"])
+                            || match_def_path(cx, did.did, &["core", "cell", "RefMut"])
+                            || match_def_path(cx, did.did, &["core", "slice", "Iter"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "Entry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "OccupiedEntry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "VacantEntry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "Iter"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "set", "Iter"]) {
+                        // Structures which are semantically similar to an &ptr.
+                        false
+                    } else if did.is_box() && in_new_function {
+                        // box in new() is okay
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyParam(param_ty) => {
+                    let ty_param_def = cx.tcx.generics_of(self.def_id).type_param(&param_ty, cx.tcx);
+                    if cx.tcx.has_attr(ty_param_def.def_id, "must_root") {
+                        ret = true;
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyRef(..) => false, // don't recurse down &ptrs
+                ty::TyRawPtr(..) => false, // don't recurse down *ptrs
+                ty::TyFnDef(..) | ty::TyFnPtr(_) => false,
+                _ => true
+            }
+        });
+        ret
+    }
 
-        fn require_rooted(cx: &LateContext, in_new_function: bool, subexpr: &hir::Expr) {
-            let ty = cx.tables.expr_ty(&subexpr);
-            if is_unrooted_ty(cx, ty, in_new_function) {
-                cx.span_lint(UNROOTED_MUST_ROOT,
-                             subexpr.span,
-                             &format!("Expression of type {:?} must be rooted", ty))
+    fn has_unrooted_generic_substs(&self, did: hir::def_id::DefId, substs: &ty::subst::Substs) -> bool {
+        let cx = self.late_cx;
+
+        let generics = cx.tcx.generics_of(did);
```

Comment:
Right, so here we want to actually have a loop, sort of like:

```
let mut did = did;
loop {
  let generics = cx.tcx.generics_of(did);
  ... // just as you have it now
  match generics.parent {
    Some(d) => { def_id = d; }
    None => { return;
  }
}
```

See also:

https://github.com/rust-lang/rust/blob/9c9424de51da41fd3d1077ac7810276f8dc746fa/src/librustc/ty/mod.rs#L771-L772

Reply:
alternatively just having the function call itself recursively would be fine